source/installer.py

```diff
@@ -622,9 +631,60 @@ def tryCopyFile(sourceFilePath,destFilePath):
 			raise OSError("Unable to copy file %s to %s, error %d"%(sourceFilePath,destFilePath,errorCode))
 
 
+_nvdaExes = {
+	"nvda.exe",
+	"nvda_noUIAccess.exe",
+	"nvda_uiAccess.exe",
+	"nvda_dmp.exe",
+	"nvda_slave.exe"
+}
+
+
+def _revertGroupDelete(tempDir: str, installDir: str):
+	"""Move all files in tempDir back to installDir, retaining the same relative path"""
+	for tempFile in pathlib.Path(tempDir).rglob("*"):
+		relativePath = tempFile.relative_to(tempDir)
+		originalPath = os.path.join(installDir, relativePath.as_posix())
+		try:
+			os.rename(tempFile.absolute(), originalPath)
+		except OSError:
+			log.exception(f"Failed to rename {tempFile} back to {originalPath}")
+
+
+def _deleteFileGroupOrFail(installDir: str, relativeFilepaths: Iterable[str]):
+	"""
+	Delete a group of files in the installer folder.
+	If any file fails to be deleted, revert the deletion of all other files.
+
+	:param installDir: an iterable of file paths relative to installDir
+	:param relativeFilepaths: an iterable of file paths relative to installDir
+
+	:raises RetriableFailure: if the files fail to be deleted.
+	"""
+	tempDir = tempfile.mkdtemp()
+	for filepath in relativeFilepaths:
+		originalPath = os.path.join(installDir, filepath)
+		if not pathlib.Path(originalPath).exists():
+			log.debug(f"Skipping remove for non-existent file: {originalPath}")
+			continue
+		tempPath = os.path.join(tempDir, filepath)
+		pathlib.Path(tempPath).parent.mkdir(parents=True, exist_ok=True)
+		try:
+			os.rename(originalPath, tempPath)

```

Comment:
My manual testing shows that while NvDA is running, I can still rename nvda.exe with no issue. However, I cannot remove it. Therefore, this rename should be replaced with a copy and then remove.